Drop CString constructors taking in a raw pointer and a length

cdumez opened the pull request:
#### a52537d513b7157f10fc5d1b9a6f9dde3053e038
<pre>
Drop CString constructors taking in a raw pointer and a length
<a href="https://bugs.webkit.org/show_bug.cgi?id=272179">https://bugs.webkit.org/show_bug.cgi?id=272179</a>

Reviewed by Darin Adler.

Drop CString constructors taking in a raw pointer and a length, in favor of the
ones taking in a std::span.

* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(StringRange::StringRange):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::inferredName const):
* Source/JavaScriptCore/parser/UnlinkedSourceCode.cpp:
(JSC::UnlinkedSourceCode::toUTF8 const):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::toCanonical):
* Source/WTF/wtf/StringHashDumpContext.h:
(WTF::StringHashDumpContext::getID):
* Source/WTF/wtf/StringPrintStream.cpp:
(WTF::StringPrintStream::toCString):
* Source/WTF/wtf/cf/FileSystemCF.cpp:
(WTF::FileSystem::fileSystemRepresentation):
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
(WTF::userVisibleString):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::CString):
(WTF::CString::init):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::latin1 const):
(WTF::String::tryGetUTF8 const):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::trimToValidUTF8Length1024):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringNullStringConstructor)):
(TEST(WTF, CStringEmptyEmptyConstructor)):
(TEST(WTF, CStringEmptyRegularConstructor)):
(TEST(WTF, CStringZeroTerminated)):
* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteUint)):
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteNegativeInteger)):

Canonical link: <a href="https://commits.webkit.org/277140@main">https://commits.webkit.org/277140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43949476c5e4ef7aa382bb5a2599b72bc862d9e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46746 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49423 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42793 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23369 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38095 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47326 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19379 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20243 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41398 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4792 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39993 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41765 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51295 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46229 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21755 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18113 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45375 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23043 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44331 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23529 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53371 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6558 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22748 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10957 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/StringPrintStream.cpp:
```diff
@@ -86,7 +86,7 @@ void StringPrintStream::vprintf(const char* format, va_list argList)
 {
     ASSERT(m_next == strlen(m_buffer));
-    return CString(m_buffer, m_next);
+    return CString({ m_buffer, m_next });
```
Can we omit `CString()` here?

darinadler replied:
Maybe not, I don’t see how it figures out to use `std::span`.

cdumez replied:
If I remember correctly, it was either:
```
return CString({ m_buffer, m_next });
```
or
```
return std::span<const char> { m_buffer, m_next };
```
(it guess because m_buffer is not `const`.

I went with the most concise version that built.

darinadler commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -66,7 +66,7 @@ CString currentExecutablePath()
     if (result == -1)
         return { };
-    return CString(readLinkBuffer, result);
+    return CString({ readLinkBuffer, static_cast<size_t>(result) });
```
Can we omit `CString()` here?

darinadler replied:
Maybe not, I don’t see how it figures out to use `std::span`.

cdumez replied:
`readLinkBuffer` is not `const` so it's the same issue as above.

darinadler commented on Source/WTF/wtf/text/CString.cpp:
```diff
@@ -45,31 +45,31 @@ Ref<CStringBuffer> CStringBuffer::createUninitialized(size_t length)
 }
 
-CString::CString(const char* str)
+CString::CString(const char* string)
```
Seems like this could call the other constructor.
```
    : CString(WTF::span(string))
```

darinadler commented on Source/WTF/wtf/text/CString.cpp:
```diff
@@ -45,31 +45,31 @@ Ref<CStringBuffer> CStringBuffer::createUninitialized(size_t length)
 }
 
-void CString::init(const char* str, size_t length)
+void CString::init(std::span<const char> string)
```
If we made one constructor call the other, we could put this into the constructor instead of having a separate function.

cdumez replied:
Good idea. I'll make the change.

cdumez committed:
Drop CString constructors taking in a raw pointer and a length
https://bugs.webkit.org/show_bug.cgi?id=272179

Reviewed by Darin Adler.

Drop CString constructors taking in a raw pointer and a length, in favor of the
ones taking in a std::span.

* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(StringRange::StringRange):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::inferredName const):
* Source/JavaScriptCore/parser/UnlinkedSourceCode.cpp:
(JSC::UnlinkedSourceCode::toUTF8 const):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::toCanonical):
* Source/WTF/wtf/StringHashDumpContext.h:
(WTF::StringHashDumpContext::getID):
* Source/WTF/wtf/StringPrintStream.cpp:
(WTF::StringPrintStream::toCString):
* Source/WTF/wtf/cf/FileSystemCF.cpp:
(WTF::FileSystem::fileSystemRepresentation):
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
(WTF::userVisibleString):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::CString):
(WTF::CString::init):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::latin1 const):
(WTF::String::tryGetUTF8 const):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::trimToValidUTF8Length1024):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringNullStringConstructor)):
(TEST(WTF, CStringEmptyEmptyConstructor)):
(TEST(WTF, CStringEmptyRegularConstructor)):
(TEST(WTF, CStringZeroTerminated)):
* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteUint)):
(TestWebKitAPI::TEST(CBORWriterTest, TestWriteNegativeInteger)):

Canonical link: https://commits.webkit.org/277140@main
